opt/optbuilder: support using table name as projection

otan opened the pull request:
* Fix projection code to be able to use the column name as a table name
  by catching undefined columns error and evaluating it as a TupleStar
* Fix scope VisitPre to do something similar to above, to make this work
  for functions.

Resolves #60549
Resolves #26719
Resolves #45854

Release note (sql change): Using the table name as a projection now
works, e.g. SELECT table_name FROM table_name or SELECT
row_to_json(table_name) FROM table_name.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64748)
<!-- Reviewable:end -->

otan commented:
I'm not sure this is the best strategy for dealing with it, but other stuff i've tried (e.g. try make tree_name_resolution recognise these) doesn't work out :\

rytaft reviewed (changes requested):
This name resolution stuff is a headache... thanks for working on this!

Does this also fix #45854?

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-:-M_0_kQ4BCnH46QMqgtd:b-gh5vek)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/project.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0fWAs5RNBzGrCYk_A:-M_0fWAt5qsVpadQ9UF1:b-6684m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/project.go#L200)):*
> ```Go
> 							panic(resolveErr)
> 						}
> 					}()
> ```

Completely ignoring the source of the panic here makes me nervous. We already have an issue open to audit the panic catching code in the optimizer since it may be catching panics that it shouldn't.

It would be great to at least put this in a helper function (e.g., `tryRunFunction`) to limit the proliferation of new panic catchers, but I wonder if it's possible to avoid the need for this completely.

For example, could you call some other function to check whether `colName` is actually a table name or something that can be resolved to a tuple?
___
*[pkg/sql/opt/optbuilder/scope.go, line 946 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0akaU4mjIKKA24MWE:-M_0akaU4mjIKKA24MWF:bbx4i54) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/scope.go#L946)):*
> ```Go
> 							Parts:    tree.NameParts{"", string(t.ColumnName)},
> 						},
> 					}
> ```

I think you could put this into a little helper function `makeTupleStarFromTableName` or something like that (and reuse in project.go).
___
*[pkg/sql/opt/optbuilder/testdata/select, line 1508 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0bRa0Bid9tetLLkUY:-M_0bRa0Bid9tetLLkUZ:bn43dpv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/testdata/select#L1508)):*
> ```
> 
> exec-ddl
> CREATE TABLE same_name (
> ```

Does the behavior you're testing here match Postgres?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> Does this also fix #45854?

yep! added a test

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-:-M_0sKVE3TqW1L4S5kq9:b-fduspi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/optbuilder/project.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0fWAs5RNBzGrCYk_A:-M_0r3aT1BSYkQXxX4p3:b-b97yvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/project.go#L200)):*
> For example, could you call some other function to check whether colName is actually a table name or something that can be resolved to a tuple?

not sure where to pull that data from :(. resolveType seems to do it.
i think the panic catch here is safe because we're always falling back to the "original" panic. it is already wrapped in a defer func, i cannot re-use this as it has a different return type to the other.
___
*[pkg/sql/opt/optbuilder/scope.go, line 946 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0akaU4mjIKKA24MWE:-M_0uZkiCTb63s_pNMkN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/scope.go#L946)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think you could put this into a little helper function `makeTupleStarFromTableName` or something like that (and reuse in project.go).
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/testdata/select, line 1508 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0bRa0Bid9tetLLkUY:-M_0rJ7H37yiEmTynEp6:bb7s9xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/testdata/select#L1508)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Does the behavior you're testing here match Postgres?
</blockquote></details>

yep!


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64748)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/project.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0fWAs5RNBzGrCYk_A:-M_0wTJy3GA1a7UgoaJj:bt1zuem) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/project.go#L200)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> For example, could you call some other function to check whether colName is actually a table name or something that can be resolved to a tuple?

not sure where to pull that data from :(. resolveType seems to do it.
i think the panic catch here is safe because we're always falling back to the "original" panic. it is already wrapped in a defer func, i cannot re-use this as it has a different return type to the other.

</blockquote></details>

I don't think it's necessarily safe, because the original panic may have been ok (e.g., originating from the optimizer code), while the new panic might have actually put the DB into a bad state (e.g.,  some type resolution code querying the DB causes a panic at the kv layer). We're not differentiating between these two cases today, but that is something I want to fix soon.

Let me poke around to see if I can suggest a better function for you to call...


<!-- Sent from Reviewable.io -->

otan committed:
opt/optbuilder: support using table name as projection

* Fix projection code to be able to use the column name as a table name
  by catching undefined columns error and evaluating it as a TupleStar
* Fix scope VisitPre to do something similar to above, to make this work
  for functions.

Release note (sql change): Using the table name as a projection now
works, e.g. SELECT table_name FROM table_name or SELECT
row_to_json(table_name) FROM table_name.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64748)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/optbuilder/project.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0fWAs5RNBzGrCYk_A:-M_2VbUw3UrIVncl6Ltr:b-62dtyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/project.go#L200)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't think it's necessarily safe, because the original panic may have been ok (e.g., originating from the optimizer code), while the new panic might have actually put the DB into a bad state (e.g.,  some type resolution code querying the DB causes a panic at the kv layer). We're not differentiating between these two cases today, but that is something I want to fix soon.

Let me poke around to see if I can suggest a better function for you to call...
</blockquote></details>

i've changed it to check for UndefinedTable, in which case I'll panic with the original error.
Otherwise, it panics with the new error.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-:-M_2bkOX5NotHGJQPTf3:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/project.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64748#-M_0fWAs5RNBzGrCYk_A:-M_2bAWT2_zCoaZ3ynzK:blf4ed1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2213c7ea62258c17c9034f49bbe6f05c78711b58/pkg/sql/opt/optbuilder/project.go#L200)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i've changed it to check for UndefinedTable, in which case I'll panic with the original error.
Otherwise, it panics with the new error.
</blockquote></details>

Thanks! Sorry I didn't get back to this before you did!


<!-- Sent from Reviewable.io -->

otan commented:
no worries, tftr!

bors r=rytaft
